File: clients/typescript/src/http.ts
/**
 * fetch-based REST client with cached, self-refreshing StreamForge auth.
 *
 * Ported from clients/python/src/streamforge/_http.py (itself a port of otc-terms'
 * lib/streamforge/server.ts sfFetch): the JWT is cached in memory for ~11h (the server issues
 * 12h tokens) and re-minted exactly once on any 401, then the request is retried once with the
 * fresh token -- if THAT also 401s, we throw rather than looping forever (a StreamForge restart
 * invalidates every token minted before it, which is a normal event, but an auth system that is
 * actually broken should fail loudly, not spin).
 */

import { AuthError } from "./errors.js";

const TOKEN_LIFETIME_MS = 11 * 60 * 60 * 1000; // server mints 12h tokens; refresh a bit early

export interface RestClientOptions {
  baseUrl: string;
  user?: string;
  password?: string;
  token?: string;
  /** false = accept self-signed/invalid TLS certs (local dev with a portless cert). */
  verify?: boolean;
}

export class RestClient {
  readonly baseUrl: string;
  private readonly user: string | undefined;
  private readonly password: string | undefined;
  private readonly verify: boolean;
  private token_: string | null;
  private tokenMintedAtMs: number | null;
  private loginInFlight: Promise<void> | null = null;

  constructor(opts: RestClientOptions) {
    this.baseUrl = opts.baseUrl.replace(/\/+$/, "");
    this.user = opts.user;
    this.password = opts.password;
    this.verify = opts.verify ?? true;
    this.token_ = opts.token ?? null;
    this.tokenMintedAtMs = opts.token ? Date.now() : null;
  }

  private fetchInit(): RequestInit {
    // Bun's fetch accepts a `tls` extension for per-request cert verification; there is no
    // portable standard fetch option for this. verify=false is spelled out by the caller
    // (connect()'s own doc comment) rather than silently defaulted, matching the design doc's
    // "a client that quietly stops verifying TLS is not a habit worth teaching" (§4).
    return this.verify ? {} : ({ tls: { rejectUnauthorized: false } } as RequestInit);
  }

  async token(): Promise<string> {
    if (this.token_ === null || this.expired()) {
      await this.login();
    }
    if (this.token_ === null) throw new AuthError("StreamForge login did not return a token");
    return this.token_;
  }

  private expired(): boolean {
    return this.tokenMintedAtMs === null || Date.now() - this.tokenMintedAtMs > TOKEN_LIFETIME_MS;
  }

  private async login(): Promise<void> {
    // Coalesce concurrent callers into one login POST rather than one per stalled request.
    if (this.loginInFlight) return this.loginInFlight;
    this.loginInFlight = this.loginNow().finally(() => {
      this.loginInFlight = null;
    });
    return this.loginInFlight;
  }

  private async loginNow(): Promise<void> {
    if (!this.user || !this.password) {
      throw new AuthError(
        "no StreamForge credentials configured -- pass user=/password= to connect(), or set " +
          "STREAMFORGE_ADMIN_USER/STREAMFORGE_ADMIN_PASS",
      );
    }
    const res = await fetch(`${this.baseUrl}/api/auth/login`, {
      ...this.fetchInit(),
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ username: this.user, password: this.password }),
    });
    if (!res.ok) {
      throw new AuthError(`StreamForge login failed: ${res.status} ${await res.text()}`);
    }
    const body = (await res.json()) as { token: string };
    this.token_ = body.token;
    this.tokenMintedAtMs = Date.now();
  }

  invalidateToken(): void {
    this.token_ = null;
  }

  /** `auth: false` skips minting/attaching a Bearer token entirely -- used for the ingest path
   * when only an ingest key is configured, so a caller that only feeds a source never forces an
   * admin login (design doc §4). */
  async request(
    method: string,
    path: string,
    init: { body?: unknown; headers?: Record<string, string>; auth?: boolean; params?: Record<string, string | number | boolean | undefined> } = {},
  ): Promise<Response> {
    const { body, headers: extraHeaders, auth = true, params } = init;
    const url = new URL(`${this.baseUrl}${path}`);
    if (params) {
      for (const [k, v] of Object.entries(params)) {
        if (v !== undefined) url.searchParams.set(k, String(v));
      }
    }
    const headers: Record<string, string> = { accept: "application/json", ...extraHeaders };
    const requestInit: RequestInit = {
      ...this.fetchInit(),
      method,
      headers,
    };
    if (body !== undefined) {
      headers["content-type"] = "application/json";
      requestInit.body = JSON.stringify(body);
    }

    if (!auth) {
      return fetch(url, requestInit);
    }

    headers.authorization = `Bearer ${await this.token()}`;
    let res = await fetch(url, requestInit);
    if (res.status === 401) {
      this.invalidateToken();
      headers.authorization = `Bearer ${await this.token()}`;
      res = await fetch(url, requestInit);
      if (res.status === 401) {
        throw new AuthError(`StreamForge rejected the re-minted token for ${method} ${path}`);
      }
    }
    return res;
  }

  get(path: string, init?: Parameters<RestClient["request"]>[2]): Promise<Response> {
    return this.request("GET", path, init);
  }

  post(path: string, init?: Parameters<RestClient["request"]>[2]): Promise<Response> {
    return this.request("POST", path, init);
  }

  delete(path: string, init?: Parameters<RestClient["request"]>[2]): Promise<Response> {
    return this.request("DELETE", path, init);
  }

  async close(): Promise<void> {
    // Nothing to release: fetch has no persistent connection object to close.
  }
}

File: clients/python/src/streamforge/_http.py
"""httpx-based REST client with cached, self-refreshing StreamForge auth.

Ported from lib/streamforge/server.ts's sfFetch: the JWT is cached in memory for ~11h (the
server issues 12h tokens) and re-minted exactly once on any 401, then the request is retried
once with the fresh token -- if THAT also 401s, we raise rather than looping forever (a
StreamForge restart invalidates every token minted before it, which is a normal event, but an
auth system that is actually broken should fail loudly, not spin).
"""

from __future__ import annotations

import threading
import time

import httpx

from .errors import AuthError

TOKEN_LIFETIME_S = 11 * 60 * 60  # server mints 12h tokens; refresh a bit early


class AuthClient:
    """Owns one httpx.Client and the login/token lifecycle for one base_url."""

    def __init__(
        self,
        base_url: str,
        user: str | None,
        password: str | None,
        *,
        verify: bool = True,
        token: str | None = None,
        timeout: float = 30.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self._user = user
        self._password = password
        self._client = httpx.Client(verify=verify, timeout=timeout)
        self._lock = threading.Lock()
        self._token = token
        self._token_minted_at = time.monotonic() if token else None

    def close(self) -> None:
        self._client.close()

    # ---- token lifecycle ----

    def token(self) -> str:
        with self._lock:
            if self._token is None or self._expired():
                self._login_locked()
            assert self._token is not None
            return self._token

    def _expired(self) -> bool:
        return self._token_minted_at is None or (time.monotonic() - self._token_minted_at) > TOKEN_LIFETIME_S

    def _login_locked(self) -> None:
        if not self._user or not self._password:
            raise AuthError(
                "no StreamForge credentials configured -- pass user=/password= to connect(), "
                "set STREAMFORGE_ADMIN_USER/STREAMFORGE_ADMIN_PASS, or add them to "
                "~/.config/streamforge/config.toml"
            )
        resp = self._client.post(
            f"{self.base_url}/api/auth/login",
            json={"username": self._user, "password": self._password},
        )
        if resp.status_code != 200:
            raise AuthError(f"StreamForge login failed: {resp.status_code} {resp.text}")
        body = resp.json()
        self._token = body["token"]
        self._token_minted_at = time.monotonic()

    def invalidate_token(self) -> None:
        with self._lock:
            self._token = None

    # ---- requests ----

    def request(self, method: str, path: str, *, auth: bool = True, **kwargs) -> httpx.Response:
        """`auth=False` skips minting/attaching a Bearer token entirely -- used for the ingest
        path when only SF_INGEST_KEY is configured, so a notebook that only feeds a source never
        forces an admin login (design doc §4)."""
        headers = dict(kwargs.pop("headers", None) or {})
        url = f"{self.base_url}{path}"
        if not auth:
            return self._client.request(method, url, headers=headers, **kwargs)

        headers["authorization"] = f"Bearer {self.token()}"
        resp = self._client.request(method, url, headers=headers, **kwargs)
        if resp.status_code == 401:
            self.invalidate_token()
            headers["authorization"] = f"Bearer {self.token()}"
            resp = self._client.request(method, url, headers=headers, **kwargs)
            if resp.status_code == 401:
                raise AuthError(f"StreamForge rejected the re-minted token for {method} {path}")
        return resp

    def get(self, path: str, **kwargs) -> httpx.Response:
        return self.request("GET", path, **kwargs)

    def post(self, path: str, **kwargs) -> httpx.Response:
        return self.request("POST", path, **kwargs)

    def delete(self, path: str, **kwargs) -> httpx.Response:
        return self.request("DELETE", path, **kwargs)

    def stream(self, method: str, path: str, **kwargs):
        """Streaming request (SSE receive) as a context manager, same auth as request()."""
        headers = dict(kwargs.pop("headers", None) or {})
        headers["authorization"] = f"Bearer {self.token()}"
        return self._client.stream(method, f"{self.base_url}{path}", headers=headers, **kwargs)
